### DISCUSSION TITLE
Rotating your application's LWA credentials

### DISCUSSION THREAD

**Discussion by moltar** [amzn/selling-partner-api-models#3501] [General] [answered]
Wouldn't that also invalidate all of the end user refresh tokens? Wouldn't that create a major PITA for all end users to have to re-authorize their accounts with all of the Amazon apps every 180 days? An average seller is probably using 5 different apps to manage their business, of which probably 4/5 have to connect with the SP API.

> As of February 6, 2023, you must rotate you Login With Amazon (LWA) credentials (client secrets) for all applications every 180 days. All credentials must be rotated by March 27, 2023. If you do not update your LWA credentials before your target rotation date, your application will lose access to SP-API. You will be notified 30 days prior to the target rotation date of LWA credentials, and you will receive periodic reminder emails as you approach your credential's target rotation date.

See for background:

- https://github.com/amzn/selling-partner-api-docs/discussions/2909
- https://developer-docs.amazon.com/sp-api/docs/rotating-your-apps-lwa-credentials

**Comment by moltar** [comment]
Ping @hudalexa

**Comment by stefnats** [comment]
I'm pretty sure, that the seller doesn't have to re-authorize your app every 3 months.

**Comment by sungolivia** [comment, marked as answer]
@stefnats  is correct you won't need to get new auths from your customers as existing refresh token will continue to work.

**Reply by jdcrutchley** [reply]
The generated APIs seem to require the client secret to generate the request.  If it changes, won't calls stop working 7 days after the rotation?  If so, that means we're back to someone having to replace credentials in every instance of every app every 180 days, right?  If not, and the refresh token is all that's used, why do the generated APIs require the client secret at all when I can get a refresh token when I create the client secret and revoke it if compromised?  I guess I'm fuzzy on what this client secret is actually doing.

**Reply by MyGuyCai** [reply]
@sungolivia  Would also like some clarification on this, do we need to rotate the LWA secret if we are using refresh tokens?

**Reply by blockberd** [reply]
I have the same concern as @jdcrutchley . Can you evaluate this @sungolivia ?

**Comment by Dragonsangel** [comment]
I can confirm that rotating the LWA token does not require new auths from our customers. We recently rotated our and everything rolled over smoothly from the old token to the new one.

One thing that did cause me concern was this paragraph in the [linked document](https://developer-docs.amazon.com/sp-api/docs/rotating-your-apps-lwa-credentials#you-rotate-your-credentials-but-continue-to-use-the-old-credentials-instead-of-the-new-credentials):

> You rotate your credentials, but continue to use the old credentials instead of the new credentials.
> You will see the following error message if you continue to use old credentials:
> {
>  "error_description": "Client authentication failed",
>  "error": "invalid_grant"
> }
> This error occurs during the Request a Login with Amazon access token step, resulting in loss of API access for the application. Use the > following procedure to resolve the issue:

From this paragraph, it was unclear if we would receive this error immediately after rotation, but before we can rollout the new token to our system. This error would come up if a new customer wants to authorize our system to access, correct?

**Comment by suofiya** [comment]
hi all，

The old LWA credentials is still valid within the expiration time of 7 days, after the new LWA credentials is rotated. It is necessary to ensure that the application runs stably and healthily after the LWA credentials is replaced. 
In other words, after replacing the new  LWA credentials, I hope to test it before the online application's configration is updated。

**Comment by ccruzv** [comment]
Hello guys! 
Does any one know if there is a non manually way to rotate and get the new client secret?

**Reply by jamesp26** [reply]
Plus 1 for this. I would rather create a Lambda that handled the rotation for us and manually remembering to do it is clearly going to cause problems.

**Reply by Akum777** [reply]
Thanks for your feedback. We are looking on ways to automate the process for our Developers (more to come on this in the coming months) - for our reference, is there an automated rotation mechanism that you prefer or are most familiar with?

**Reply by moltar** [reply]
AWS Secrets Manager rotation Lambda

https://docs.aws.amazon.com/secretsmanager/latest/userguide/rotating-secrets.html

**Reply by owenashurst** [reply]
> AWS Secrets Manager rotation Lambda
> 
> https://docs.aws.amazon.com/secretsmanager/latest/userguide/rotating-secrets.html

This depends on whether you can programmatically call some Sp-API to rotate the credentials. Is there an API call for this? If so, Secrets Manager is a good option.

**Reply by moltar** [reply]
@owenashurst This was in response to @Akum777 asking what mechanism is preferred. There is no known way to do this now. However, I am just stating my preferred mechanism for rotation, in case they want to add support for it.

**Reply by Akum777** [reply]
Thanks @moltar.

**Reply by ZaidPathan** [reply]
is there any update on this? @Akum777

**Reply by Akum777** [reply]
Hi @ZaidPathan - apologies, no updates on it yet. I can follow up by end of August some details of timeline and plans.

**Comment by rugved1991** [comment]
Hi,

You can refer to this blog for more information on how to programmatically rotate LWA credentials using AWS.
https://developer-docs.amazon.com/sp-api-blog/docs/rotate-your-sp-api-credentials-using-aws

Best,
Rugved
Solutions Architect, SP-API
